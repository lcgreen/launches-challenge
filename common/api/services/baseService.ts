export type SchemaType = "launches" | "payloads";

export type ApiQueryParams = {
  options: {
    limit: number;
    page: number;
    sort?: Record<string,string>;
  };
};

export type BaseServiceResponse<Doc> = {
  totalDocs: number;
  offset: number;
  limit: number;
  totalPages: number;
  page: number;
  pagingCounter: number;
  hasPrevPage: boolean;
  hasNextPage: boolean;
  prevPage: null | number;
  nextPage: null | number;
  docs: Doc[];
};


const buildUri = (path: SchemaType, version: string, type: string = '') => {
  return `${process.env.API_URL}/${version}/${path}/${type}`;
}

export default {
  /**
   * Queries the API
   * @param {string} path query path
   * @param {object} args query arguments
   * @returns {object} json data
   */
  query: async <Doc>(
    schema: SchemaType,
    version: 'v4' | 'v5',
    args: ApiQueryParams
  ): Promise<BaseServiceResponse<Doc>> => {
    const uri = buildUri(schema, version, 'query');
    const response = await fetch(uri, {
      method: 'POST',
      // body: {
      //   options: args.options,
      //   query: {
      //     upcoming: true
      //   }
      // }
    });

    if (!response.ok) {
      throw new Error(response.statusText);
    }

    return await response.json();
  },

  all: async <Doc>(
    schema: SchemaType,
    version: 'v4' | 'v5',
  ): Promise<Doc[]> => {
    const uri = buildUri(schema, version);
    const response = await fetch(uri, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(response.statusText);
    }

    return await response.json();
  },
};
