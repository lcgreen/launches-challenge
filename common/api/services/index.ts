export { default as LaunchesService } from "./launches";
export { default as PayloadService } from './payloads';