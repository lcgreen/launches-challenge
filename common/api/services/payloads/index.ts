import PayloadDocument from "@@/api/schema/v4/payloadDocument";
import BaseService  from "../baseService";

export type Payload = {
  id: string;
  type: string;
}

const mapPayloadType = (document: PayloadDocument): Payload => {
  return {
    id: document.id,
    type: document.type,
  };
};

export default {
  all: async () => {
    const resp = await BaseService.all<PayloadDocument>('payloads', 'v4');
    if (resp.length) {
      return resp.map((doc) => mapPayloadType(doc));
    }
  },
};
