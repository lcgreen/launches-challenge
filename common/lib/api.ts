export const fetcher = async (url) => {
  const response = await fetch(url).then((res) => res.json());
  if (!response.ok) {
    throw new Error('Failed to fetch API');
  }
  return response;
};
