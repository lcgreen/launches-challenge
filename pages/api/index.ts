import { NextApiHandler } from "next";

const handler: NextApiHandler = async (_req, res) => {
  res.status(200).json({ api: ['launches', 'payloads'] });
};

export default handler;