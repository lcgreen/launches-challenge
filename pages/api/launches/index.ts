import { NextApiHandler } from 'next';
import { LaunchesService } from '@@/api/services';

const handler: NextApiHandler = async (req, res) => {
  try {
    // query the launches api
    const json = await LaunchesService.all();
    res.status(200).json({
      ok: true,
      body: json,
    });
  } catch (error) {
    res.status(500).json({
      ok: false,
      error: error.message,
    });
  }
};

export default handler;
