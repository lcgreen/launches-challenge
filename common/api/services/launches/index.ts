import PayloadDocument from "@@/api/schema/v4/payloadDocument";
import LaunchDocument from "@@/api/schema/v5/launchDocument";
import { PayloadService } from "..";
import BaseService from "../baseService";
import { Payload } from "../payloads";

export type Launch = {
  id: string
  flight_number: number
  name: string
  date_utc: string
  serial: string
  payload_id: string
  image: string
  success: boolean
  failures?: {
    time: number
    altitude: number | null
    reason: string
  }[]
  payload?: Payload
}

const mapLaunchType = (document: LaunchDocument, payloads: PayloadDocument[]): Launch => {
  return {
    id: document.id,
    name: document.name,
    date_utc: document.date_utc,
    serial: document.cores[0].core,
    image: document.links.patch.small,
    success: document.success,
    failures: document?.failures ?? [],
    payload_id: document?.payloads?.[0] ?? '',
    flight_number: document.flight_number,
    payload: payloads.find((payload) => payload.id === document?.payloads?.[0]) ?? null,
  };
};

export default { 
  all: async () => {
    const resp = await BaseService.all<LaunchDocument>('launches', 'v5');
    const payloads = await PayloadService.all();
    if (resp.length) {
      const sortedLaunches = resp.sort((a, b) => a.flight_number - b.flight_number).slice(0, 10);
      return sortedLaunches.map((doc) => mapLaunchType(doc, payloads));
    }
  },
  mapLaunchType,
}