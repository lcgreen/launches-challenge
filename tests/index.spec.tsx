import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Launches from '../pages';
import LaunchDocument from '@@/api/schema/v5/launchDocument';
import { LaunchesService } from '@@/api/services';

let data = [];
beforeEach(async () => {
  const launches = import('@@/tests/fixtures/launches.json');
  const payloads = import('@@/tests/fixtures/payloads.json');

  const payloadData = (await payloads).body.docs;
  const launchData = (await launches).body.docs;

  const sortedLaunches = launchData.sort((a, b) => a.flight_number - b.flight_number).slice(0, 1);
  data = sortedLaunches.map((doc: LaunchDocument) => LaunchesService.mapLaunchType(doc, payloadData));
});

jest.mock('./../hooks/useLaunches', () => () => ({
  loading: false,
  isError: false,
  data: {
    body: [...data],
  },
}));

describe('Launches', () => {
  it('renders the component', () => {
    const { container } = render(<Launches />);
    expect(container).toMatchSnapshot();
  });

  it('renders the modal when card is clicked', () => {
    render(<Launches />);
    fireEvent.click(screen.getByTestId('card-5eb87cd9ffd86e000604b32a'));
    expect(
      screen.getByRole('heading', {
        name: 'Payload Information for FalconSat',
      })
    ).not.toBeNull();
  });

  it('displays a table displaying the payload information', () => {
    render(<Launches />);
    fireEvent.click(screen.getByTestId('card-5eb87cd9ffd86e000604b32a'));
    expect(
      screen.getByRole('table')
    ).not.toBeNull();
  })
});
