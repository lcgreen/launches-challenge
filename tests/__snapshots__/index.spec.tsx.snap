// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launches renders the component 1`] = `
<div>
  <main>
    <h1
      class="header"
    >
      Launches
    </h1>
    <div
      class="container"
    >
      <div
        data-testid="card-5eb87cd9ffd86e000604b32a"
      >
        <div
          class="container"
          tabindex="0"
        >
          <div
            class="header"
          >
            <div
              class="container"
            >
              <img
                alt="FalconSat"
                class="image"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=96&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
              <div
                class="body"
              >
                <div
                  class="title"
                >
                  FalconSat
                </div>
                <span
                  class="date"
                >
                  Fri Mar 24 2006
                </span>
              </div>
            </div>
          </div>
          <div
            class="body"
          >
            <ul
              class="list"
            >
              <li>
                <div
                  class="status"
                >
                  <svg
                    class="icon-error"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.1"
                    viewBox="0 0 16 16"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M15.854 12.854c-0-0-0-0-0-0l-4.854-4.854 4.854-4.854c0-0 0-0 0-0 0.052-0.052 0.090-0.113 0.114-0.178 0.066-0.178 0.028-0.386-0.114-0.529l-2.293-2.293c-0.143-0.143-0.351-0.181-0.529-0.114-0.065 0.024-0.126 0.062-0.178 0.114 0 0-0 0-0 0l-4.854 4.854-4.854-4.854c-0-0-0-0-0-0-0.052-0.052-0.113-0.090-0.178-0.114-0.178-0.066-0.386-0.029-0.529 0.114l-2.293 2.293c-0.143 0.143-0.181 0.351-0.114 0.529 0.024 0.065 0.062 0.126 0.114 0.178 0 0 0 0 0 0l4.854 4.854-4.854 4.854c-0 0-0 0-0 0-0.052 0.052-0.090 0.113-0.114 0.178-0.066 0.178-0.029 0.386 0.114 0.529l2.293 2.293c0.143 0.143 0.351 0.181 0.529 0.114 0.065-0.024 0.126-0.062 0.178-0.114 0-0 0-0 0-0l4.854-4.854 4.854 4.854c0 0 0 0 0 0 0.052 0.052 0.113 0.090 0.178 0.114 0.178 0.066 0.386 0.029 0.529-0.114l2.293-2.293c0.143-0.143 0.181-0.351 0.114-0.529-0.024-0.065-0.062-0.126-0.114-0.178z"
                    />
                  </svg>
                  <div>
                    Failure: 
                    merlin engine failure
                  </div>
                </div>
              </li>
              <li>
                Serial: 
                5e9e289df35918033d3b2623
              </li>
              <li>
                Payload Id: 
                5eb0e4b5b6c3bb0006eeb1e1
              </li>
              <li>
                Payload Type: 
                Satellite
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="overlay"
      style="display: none;"
    />
  </main>
</div>
`;
